### DISCUSSION TITLE
Arabic Language Support / RTL text

### DISCUSSION THREAD

**Discussion by MartinThoma** [py-pdf/pypdf#1547] [General]
A user at https://stackoverflow.com/q/75050321/562769 offered to help us with improving the support for Arabic. Are there any specific cases where we were uncertain? Anything specific we can ask?

**Comment by pubpub-zz** [comment]
Any help would be welcomed. Did he notices some issues ?

**Reply by karenlmcneil** [reply]
No, she did not. ;-p
But I haven't tried pypdf specifically. But every other pdf tool I've tried generally has problems with Arabic, so I would be surprised if you didn't. I have plenty of Arabic PDFs, and would be happy to help. Probably better to email me (since I don't check these sites very often — that's why it took me so long to notice Martin's response). My email address is here: http://www.tunisiya.org/team/

**Reply by MartinThoma** [reply]
Please also let us know if you (or anybody) notice issues. We don't use Arabic, so we will neither encounter issues nor be able to tell if there is an issue at all.

Up to my knowledge, pypdf should work with Arabic :-)

**Reply by karenlmcneil** [reply]
Ok, now you got me curious ...

![Screen Shot 2023-01-23 at 4 50 21 PM](https://user-images.githubusercontent.com/47640283/214158035-1d4280da-917c-40b8-a183-4fe2decaaa08.png)
![Screen Shot 2023-01-23 at 4 51 22 PM](https://user-images.githubusercontent.com/47640283/214158050-aba15785-07fe-419c-8437-c4b5633aa6b8.png)

Well, it's better than some others I've seen. The text isn't backwards (yay!), but some of the words are still scrambled. The word that begins the first line (reading from the right) shows up towards the end of the line in the extracted text. And several of the words on that first line are all run together for some reason. It might be the diacritics, that screws a lot of systems up.

It also reverses the order of letters that are stacked on top of eachother (this is a sample from a novel and it uses a pretty font).

So, yeah, you would probably need to do some work on it before it can usefully support Arabic. But good on you for the text at least not being backwards!

**Reply by karenlmcneil** [reply]
To give you a better idea, of the 35 words in these two paragraphs, pypdf got 20 of them right. It seems that it correctly read all the words that do not have a diacritic and that don't have any stacked letters. Although these 20 words are not necessarily in the right order. But most of them are.

**Reply by pubpub-zz** [reply]
@karenlmcneil
can you provide the pdf that shows this behavior ?

**Reply by karenlmcneil** [reply]
Sure: 

[arabic_sample.pdf](https://github.com/py-pdf/pypdf/files/10485679/arabic_sample.pdf)

**Reply by MartinThoma** [reply]
Thank you already for your help @karenlmcneil :hugs: 
If you want, I can add you as a contributor: https://pypdf.readthedocs.io/en/latest/meta/CONTRIBUTORS.html :-) Just let me know which name (McNeil, Karen?) and which link (e.g. https://github.com/karenlmcneil ? ) I should add :-)

**Reply by karenlmcneil** [reply]
Sure! Although I guess that means I'll have to actually add something to my Github profile ... And, yes, that name and link are correct

**Reply by karenlmcneil** [reply]
If you need anything else, send me an email. (I don't use this site very often, so I might not see a message for a long time.)
